internal/sqlsmith: do not generate collated string return types

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11152646?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11152646?buildTab=artifacts#/) on master @ [8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6](https://github.com/cockroachdb/cockroach/commits/8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6):


```
	defaultdb.public.seed AS tab145,
	defaultdb.public.seed AS tab146,
	defaultdb.public.seed AS "'t ab147",
	defaultdb.public.seed AS "ta\\U00026A09b148";

STMT: 80
DELETE FROM
	defaultdb.public.seed AS tab149
USING
	defaultdb.public.seed AS tab150, defaultdb.public.seed AS tab151
WHERE
	false;

STMT: 81
INSERT
INTO
	defaultdb.public.seed AS 😡tab152 (_int2, _int8, _float4, _date, _bool, _decimal, _string, _bytes, _jsonb, _enum)
VALUES
	(
		NULL,
		(-5553610400454300411):::INT8,
		(-1.1833046674728394):::FLOAT8,
		'1987-01-11':::DATE,
		true,
		(-2.979700723675381031E+26):::DECIMAL,
		e'U\\J\x01\x1b]J\x1aI':::STRING,
		e'\\x58':::BYTES,
		'{"bar": null, "i8RwPJ8K": true, "iD1H": 2.1318145374221587}':::JSONB,
		'howdy':::greeting
	);

STMT: 82
UPDATE
	defaultdb.public.seed AS "tAb153"
SET
	_int4 = tab154._int8
FROM
	defaultdb.public.seed AS tab154, defaultdb.public.seed AS "Tab155"
WHERE
	true;

    sqlsmith_test.go:201: CREATE FUNCTION func2(IN p0 TSQUERY, IN p1 TSVECTOR, IN p2 BIT(36)[], IN p3 RECORD, IN p4 REGPROCEDURE, IN p5 BIT(8), IN p6 OID)
        	RETURNS STRING COLLATE en_US
        	RETURNS NULL ON NULL INPUT
        	NOT LEAKPROOF
        	LANGUAGE SQL
        	AS $funcbody$INSERT INTO defaultdb.public.seed AS "tab 156" SELECT int2((-1.3790116553859837):::FLOAT8::FLOAT8)::INT2 AS c🙃ol209, 146801082:::INT8 AS col210, (-8755261623632097436):::INT8 AS col211, (-0.8992149829864502):::FLOAT8 AS "col'212", NULL AS col213, '1988-06-09':::DATE AS col214, '2014-09-09 01:06:31.000191':::TIMESTAMP AS col215, '2024-05-12 04:47:34.000549+00':::TIMESTAMPTZ AS col216, '5 years 1 mon 644 days 21:10:02.93671':::INTERVAL AS col😤217, true AS col218, NULL AS "c%vol219", '':::STRING AS col220, '\x44':::BYTES AS col221, '49672ab0-59a3-4cd1-b915-ccdf673d10f9':::UUID AS col222, '117.27.81.171/2':::INET AS "co l223", '{"c": [false], "ptaH": null, "vpJNjXZx": {"DFRc": null, "c": null}}':::JSONB AS "c%vol224", 'hi':::greeting AS "co'l225" RETURNING e'\U000DEA2C' COLLATE en_US AS "r͌et 4"$funcbody$: at or near "collate": syntax error
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/24a1830a60e8c96e6a05506359144db7/logTestGenerateParse3710124245
--- FAIL: TestGenerateParse (13.01s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30284

--- a/pkg/internal/sqlsmith/relational.go
+++ b/pkg/internal/sqlsmith/relational.go
@@ -1043,7 +1043,7 @@ func (s *Smither) makeCreateFunc() (cf *tree.CreateRoutine, ok bool) {
 				// If the rtype isn't a RECORD, change it to rrefs or RECORD depending
 				// how many columns there are to avoid return type mismatch errors.
 				if rtyp != types.AnyTuple {
-					if len(rrefs) == 1 && s.coin() {
+					if len(rrefs) == 1 && s.coin() && rrefs[0].typ.Family() != types.CollatedStringFamily {
 						rtyp = rrefs[0].typ
 					} else {
 						rtyp = types.AnyTuple

